### PR TITLE
 Store secrets in node in tmpfs

### PR DESCRIPTION
**Pull request by pmorie**
WIP for #4602

### CODE REVIEW

**Comment by pmorie** [comment]
@googlebot check your facts, bot

**Comment by thockin** [comment]
I was contemplating banging out tmpfs as a volume, but I took a very different tack.  Maybe there's use for both approaches.

I want to offer tmpfs as a volume type and, rather than a big global tmpfs which requires all this noise, actually just mount it directly per-pod.

It sounds like maybe we need a private-for-kubelet tmpfs, too, which we can expose as secrets without the user knowing it's tmpfs?   Can we reduce this PR to that?

**Comment by pmorie** [comment]
@thockin I think per-pod tmpfs volumes would be compatible with this - the tmpfs volume plugin could just do a tmpfs mount at the local storage pod volume dir

> It sounds like maybe we need a private-for-kubelet tmpfs, too, which we can expose as secrets without the user knowing it's tmpfs? Can we reduce this PR to that?

Not quite sure about the 'which we can expose as secrets' wording here.

**Comment by pmorie** [comment]
@thockin One thing @erictune mentioned in connection with per-pod tmpfs mounts is an upper limit (1024) under certain kernels on the number of tmpfs mounts.  I spent a few minutes looking for information about that but haven't been able to find anything yet.

**Comment by thockin** [comment]
I'm perfectly happy with a 1024 limit for a long time :)  Maybe we limit it
to 2 per pod or something, but otherwise that will last a while.

**Comment by thockin** [comment]
I'll circle back to this after reading the secrets threads.  Sorry being so
far under water :)

**Comment by pmorie** [comment]
@thockin I realized looking at the title of the pull request that it might be a little misleading.  This PR is to make the secret volume plugin store things in tmpfs so as not to allow secrets to come to rest on the node, not to introduce a tmpfs volume type.

**Comment by thockin** [comment]
I responded on the wrong thread - please see my response on the issue.
Let's discuss there then come back to the PR.

**Comment by erictune** [comment]
@thockin 
I think either per-pod or per-node tmpfs can work.  Let's build a list of pros and cons together. I'll start.
  - rotating disks can't easily be filesystem-per-pod.  Perhaps some advantage to consistency across media types. Result: advantage per-node
  - whatever mechanism we use for resource isolation on rotating disks with filesystems should also work for tmpfs with filesystems.  Result: tie
  - I can't see there being much performance difference.  Result: tie

**Comment by pmorie** [comment]
@erictune @thockin PTAL; there are some remaining TODOs but I believe the functionality is there.

**Comment by pmorie** [comment]
@thockin I'm particularly interested in thoughts you might have about the use of the `mount.Mounter` interface in the kubelet

**Comment by pmorie** [comment]
E2E passes locally; going to do a run in GCE now

**Comment by pmorie** [comment]
Looks like there's a need to be able to use the kubelet from a non-root context for tests; I will add that in more thoroughly.

**Comment by pmorie** [comment]
@thockin @erictune code review plz

**Comment by pmorie** [comment]
Full disclosure: there are still some TODOs and cleanup needed but the main ideas are all there.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -326,6 +372,20 @@ func dirExists(path string) bool {
 }
 
 func (kl *Kubelet) setupDataDirs() error {
+	err := kl.setupLocalDataDirs()
+	if err != nil {
+		return err
+	}
+
+	err = kl.setupTmpfsDataDirs()
```
Comment (liggitt) [inline pkg/kubelet/kubelet.go]: feels weird for this to be called even when mountTmpfs is false

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -339,6 +399,32 @@ func (kl *Kubelet) setupDataDirs() error {
...
+	// FUTURE: determine mount size from kubelet property
+	// TODO: determine the right way to encapsulate concerns for test-only kubelet (ie, non-root)
+	if kl.mountTmpfs {
+		label := "mode=0755,size=10g"
+		err = kl.mounter.Mount("tmpfs", kl.tmpfsRootDirectory, "tmpfs", uintptr(tmpfsMountFlags), label)
+		if err != nil {
+			return fmt.Errorf("error mounting tmpfs root directory (%v): %v", kl.tmpfsRootDirectory, err)
+		}
+	}
+	if err = os.MkdirAll(kl.getTmpfsPodsDir(), 0750); err != nil {
```
Comment (liggitt) [inline pkg/kubelet/kubelet.go]: should these get called when mountTmpfs is false?

**Comment by pmorie** [comment]
@erictune @thockin I know you're both busy this week but bumping anyway

**Comment by erictune** [comment]
@thockin i can pick this up if you want.

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -160,8 +163,6 @@ func (s *KubeletServer) Run(_ []string) error {
 		glog.Warningf("No API client: %v", err)
 	}
 
-	glog.Infof("Using root directory: %v", s.RootDirectory)
```
Comment (thockin) [inline cmd/kubelet/app/server.go]: why you killed my log?

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -108,6 +110,9 @@ func NewMainKubelet(
 		etcdClient:                     etcdClient,
 		kubeClient:                     kubeClient,
 		rootDirectory:                  rootDirectory,
+		tmpfsRootDirectory:             path.Join(rootDirectory, "tmpfs"), // TODO: pass in from cli
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: Can you capture this in a function alongside getPluginsDir() and getPodsDir() ?  That way - all of the root-dir schema is roughly in the same area.

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -317,6 +326,43 @@ func (kl *Kubelet) getPodContainerDir(podUID types.UID, ctrName string) string {
 	return newPath
 }
 
+// getTmpfsRootDir returns the full path to the kubelet's tmpfs root directory.
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: I'm not sure we need all of these functions, nor do we really need the full matrix of storage media * reason.  If we agree where we want to go with it (some sort of characteristic matching), I can live with a purpose-built interface here.  Maybe just keep getTmpfsPodVolumeDir() and expose it as Host.GetTmpfsPodVolumeDir() with a TODO that it should become a more general interface.

That should buy time until we REALLY have to think about it, and still be pretty easy to convert.

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -339,6 +399,32 @@ func (kl *Kubelet) setupDataDirs() error {
 	return nil
 }
 
+func (kl *Kubelet) setupTmpfsDataDirs() error {
+	kl.tmpfsRootDirectory = path.Clean(kl.tmpfsRootDirectory)
+	// TODO: handle edge cases like mount already exists
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: Yeah this TODO is a big one.  We have an "isMountPoint" function floating about somewhere.  Maybe a good proxy.  Or actually process /proc/mounts (we have a function for that too in mount.Interface)

File: pkg/kubelet/tmpfs_linux.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"syscall"
+)
+
+const tmpfsMountFlags = syscall.MS_NOEXEC | syscall.MS_NOSUID | syscall.MS_NODEV
```
Comment (thockin) [inline pkg/kubelet/tmpfs_linux.go]: Why not cast to uint here?

File: pkg/kubelet/kubelet_test.go [hunk 8]
```diff
@@ -61,6 +62,8 @@ func newTestKubelet(t *testing.T) (*Kubelet, *dockertools.FakeDockerClient) {
 		t.Fatalf("can't make a temp rootdir: %v", err)
 	} else {
 		kubelet.rootDirectory = tempDir
+		kubelet.tmpfsRootDirectory = path.Join(tempDir, "tmpfs")
```
Comment (thockin) [inline pkg/kubelet/kubelet_test.go]: Can you test that the dir actually gets made?

File: pkg/kubelet/tmpfs_other.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
+// +build !linux
+
+/*
+Copyright 2014 Google Inc. All rights reserved.
```
Comment (thockin) [inline pkg/kubelet/tmpfs_other.go]: 2015 on all

**Comment by thockin** [comment]
Overall I am OK with this, but I think we can drop some of it until we have a full proposal for tmpfs.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (pmorie) [inline pkg/kubelet/kubelet.go]: Will do; I agree that's better.

File: cmd/kubelet/app/server.go (hunk 3 above)
Comment (pmorie) [inline cmd/kubelet/app/server.go]: @thockin Heh, I had left this in from an earlier secrets PR and was annoyed at myself for letting my debug stuff slip in.  I'll add it back.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (pmorie) [inline pkg/kubelet/kubelet.go]: Rereading this again, I found some ambiguity; did you mean that you didn't think tmpfsRootDirectory needed to be a field in the Kubelet, and that it should just be a function that joined to the root dir?

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (pmorie) [inline pkg/kubelet/kubelet.go]: I don't think we need them; i'll revert the changes for most of this.

File: pkg/kubelet/kubelet.go (hunk 6 above)
Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @thockin I've been looking at the pattern gce_pd uses for this; it tries to unmount the directory if it's already a mountpoint, and remount it.  The `isMountPoint` function is in the gce_pd plugin; the build flags are a little weird; gce_pd uses `!windows` as the flag for the code with unix-mount awareness; the `pkg/util/mount` package uses `linux` as the build flag.  I think this code is closely related enough that it should probably all just go into `pkg/util/mount`, but we need to reconcile the build flags first.

All that said - should we try to unmount an existing mountpoint at the tmpfs storage root?  If we make the tmpfs root purely a function of the root dir (ie, don't allow you to specify some arbitrary mountpoint for tmpfs), it should be fine to remount.  If we allow you to specify an arbitrary path, it might be dangerous to do this.

Personally I think we should:

1.  Make the tmpfs root a function of the root dir for the kubelet
2.  Unmount / remount an existing tmpfs mount.

Thoughts?

Comment (markturansky) [inline pkg/kubelet/kubelet.go]: @pmorie There is duplication between gce_pd and the new proposed plugins for NFS and iCSCI, specifically around methods like ```isMountPoint```.    A refactoring in this area would help those other plugin authors.

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @thockin @markturansky I think `linux` would be a fine build-flag for the mount stuff personally.  Unless I'm mistaken I don't think we have any !linuxes in mind for platforms to run a kubelet on.

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: After talking to @markturansky I'm inclined to break out the mount refactoring into its own PR and rebase onto it.  Objections, @thockin?

**Comment by pmorie** [comment]
@thockin I am going to refactor the mount code in a separate PR and rebase onto it for ease of review.

File: pkg/kubelet/kubelet.go (hunk 6 above)
Comment (markturansky) [inline pkg/kubelet/kubelet.go]: Nitpicking the ```linux``` flag above ... I run Kubelet on OSX, but it's pods run in a boot2docker VM.  Not sure if that affects into what you are saying.

Comment (thockin) [inline pkg/kubelet/kubelet.go]: The reason we changed to !windows was OSX and unit tests, if I recall.

If the tmpfs is a function of the root directory we can't blindly unmount/remount it - it will have been bind-mounted into pods, so either we won't be allowed to unmount it or we will totally screw over any running pods.  This is important - kubelet MUST be able to crash and restart without killing pods.

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: Good point, @thockin

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @thockin 

> If the tmpfs is a function of the root directory we can't blindly unmount/remount it - it will have been bind-mounted into pods, so either we won't be allowed to unmount it or we will totally screw over any running pods. This is important - kubelet MUST be able to crash and restart without killing pods.

I think this holds no matter what the path of the tmpfs root is, and therefore the best thing to do when the mountpoint already exists is to just skip the mount attempt.  Considering it to be an error isn't an option - the mount point already existing is expected.  Eventually the kubelet will crash or go down for maintenance, and when it comes back up, the mountpoint will already exist.

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @thockin mount refactor in #5036

**Comment by pmorie** [comment]
@eparis PTAL

**Comment by pmorie** [comment]
I'm going to split this up into a couple of different PRs.   This PR is going to be rescoped to just the tmpfs changes; usage in secret volume plugin will be split into a separate PR.

**Comment by pmorie** [comment]
@thockin I've added in some basic selinux treatment but there are some edge cases to code for.  Going to do a clean e2e run and establish that everything still works after slicing and dicing / rebase.

**Comment by pmorie** [comment]
Going to split off new PRs for tmpfs and the changes to the secret volume plugin.
